### PR TITLE
Bump cppcheck action to run on Ubuntu 22.04

### PR DESCRIPTION
**Pull request by turol**
To see which cppcheck version it's running and what breaks

### CODE REVIEW

**Comment by fragglet** [comment]
Is the failed check expected? (I assume it's depending on the other PRs)

**Comment by turol** [comment]
Yes, the failure was expected.

**Comment by turol** [comment]
One of the failures is this:

```
src/doom/g_game.c:315:30: error: Array 'weapon_order_table[9]' accessed at index 9, which is out of bounds. [arrayIndexOutOfBounds]
    return weapon_order_table[i].weapon_num;
                             ^
src/doom/g_game.c:299:16: note: Assuming that condition 'i<(sizeof(weapon_order_table)/sizeof(*weapon_order_table))' is not redundant
    for (i=0; i<arrlen(weapon_order_table); ++i)
               ^
```
Link to code https://github.com/chocolate-doom/chocolate-doom/blob/88f5ae5be28d3c472c6d4e060304b877d52ec139/src/doom/g_game.c#L315

I think this can't happen because `weapon` must always be in `weapon_order_table` but cppcheck does not know this. Does anyone have ideas how to make it shut up? Can we use `assert`?  @mfrancis95 @fabiangreffrath @fragglet

**Comment by fabiangreffrath** [comment]
This is my idea for the upcoming Woof patch:
https://github.com/fabiangreffrath/woof/pull/771/files#diff-5ac47823857d7eb6d0f0288efcdd867b37fec15472c05c15dc94f5ef2ed57cfb

**Comment by turol** [comment]
Well that makes cppcheck shut up but it's kind of redundant. That condition should never be hit unless something is seriously wrong. I'd prefer something that compiles to no-op on release like `assert`. And if we end up adding an explicit `if` there should be either a log message or a fatal error.

**Comment by fabiangreffrath** [comment]
Well, most of the changes triggered by cppcheck are to make cppcheck shut up. 😉 Alternatively we could add something like `i = i % arrlen(weapon_order_table)` if you don't like another conditional.

**Comment by turol** [comment]
Adding 
```
    // the current weapon must be present in weapon_order_table
    // otherwise something has gone terribly wrong
    assert(i < arrlen(weapon_order_table));
```
make cppcheck shut up. What's our policy about assertions? There's a few in common code but not too many.

**Comment by turol** [comment]
Fixed cppcheck warnings. It now passes but the linter seems to have broken. I did not touch that in this PR...

File: src/doom/g_game.c [hunk 1]
```diff
@@ -304,6 +305,10 @@ static int G_NextWeapon(int direction)
         }
     }
 
+    // The current weapon must be present in weapon_order_table
+    // otherwise something has gone terribly wrong
+    assert(i < arrlen(weapon_order_table));
```
Comment (fabiangreffrath) [inline src/doom/g_game.c:310]: That'd be the only place where we use `assert()` in actual game code. Not sure about it. @fragglet ?

Comment (turol) [reply src/doom/g_game.c:310]: That depends on how we define "game code". There are already asserts in `i_main.c`, `i_sdlsound.c`, `m_config.c`, `midifile.c` and the OPL code.

**Comment by turol** [comment]
@fragglet ping

**Comment by turol** [comment]
@fragglet  Your input is required to resolve a difference of opinion regarding assert

**Comment by turol** [comment]
@fragglet Happy new year. Your input is still required on this issue.

**Comment by fabiangreffrath** [comment]
If we could get rid of the ugly `assert()`, I think we could merge this. I already proposed an alternative solution here:

https://github.com/fabiangreffrath/woof/commit/f70d372907e300de55d200d158e050bab494e678#diff-5ac47823857d7eb6d0f0288efcdd867b37fec15472c05c15dc94f5ef2ed57cfb

**Comment by turol** [comment]
I'm not sure that's a safe change. There appears to be only one caller for this function
https://github.com/chocolate-doom/chocolate-doom/blob/b00157f19fee71bfdce004092dc0a75ef7b94ffe/src/doom/g_game.c#L499
and it has no special handling for `wp_nochange`. Instead it appears to be put into a ticcmd which goes directly to network and/or demo (I think). Now this should never happen but if it does I'm not sure what happens.

I disagree on `assert` being ugly. It's meant specifically for things which should not happen.

Since we can't agree and The Boss doesn't seem to care, would you accept if I pinged other maintainers and we voted on this?

**Comment by fabiangreffrath** [comment]
The point is that the case we catch here actually *never happens*. We just do this to silence cppcheck.

> Since we can't agree and The Boss doesn't seem to care, would you accept if I pinged other maintainers and we voted on this?

Sure, why not.

**Comment by turol** [comment]
@mfrancis95 @jmtd @mikeday0 @rfomin @ceski-1 @JNechaevsky Please give on your opinion on what we should do.

**Comment by rfomin** [comment]
> That depends on how we define "game code". There are already asserts in i_main.c, i_sdlsound.c, m_config.c, midifile.c and the OPL code.

They are all new, there was no `assert()` in the original code. If we start using it, how about getting rid of all the `RANGECHECK` and replacing them with `assert()`? Also, how about `SDL_assert()`? 

I vote for removing cppcheck, it has never been useful. If that is out of the question, the original approach is checking with the `RANGECHECK` macro.

**Comment by JNechaevsky** [comment]
I'm afraid I can't suggest anything useful in this case. Despite of rarely using [GUI version](https://cppcheck.sourceforge.io/) of Cppcheck for own needs, I mostly rely on `-Wall` `-Wdeclaration-after-statement` `-Wredundant-decls` and `-Wpedantic` compiler options just to make sure that build log is always 100% clean under MSYS's GCC and Clang.

Thus, I can only recommend to hear our respected colleagues here.

**Comment by fabiangreffrath** [comment]
I find the `RANGECHECK` approach reasonable. If `i` ever reaches `arrlen(weapon_order_table)` we call `I_Error()` and return. And `RANGECHECK` should be defined during CI anyway.

**Comment by mikeday0** [comment]
I too like the idea of using `RANGECHECK` and `I_Error()` here. I have nothing against `assert()`, but I prefer sticking to the existing idiom in the codebase. Maybe `RANGECHECK` could eventually be replaced by `assert()` in a future clean-up?

**Comment by ceski-1** [comment]
I did a brief review and there seems to be no examples of `RANGECHECK` being used if it wasn't already present in the original code. But `RANGECHECK` just leads to `I_Error()` and there are plenty of calls to that function that are unique to Chocolate, with only a few instances of `assert()`. If this is a for an error that should never happen that seems like the use case for `assert()`. But if that's not preferred, then call `I_Error()` directly and skip the `RANGECHECK` stuff. I don't have a strong opinion either way.

**Comment by ceski-1** [comment]
> I vote for removing cppcheck, it has never been useful. 

This is probably the big picture question here.

**Comment by fabiangreffrath** [comment]
Alright, how about this then:

```c
if (i >= arrlen(weapon_order_table))
{
    I_Error("Current weapon not present in weapon_order_table");
}
```

`I_Error()` is marked `NORETURN` in `src/i_system.h` and cppcheck should understand this.

**Comment by jmtd** [comment]
Thanks for soliciting my opinion but it's probably worthless ;) I value code clarity for humans over other considerations. Something that compiles away entirely is a bonus.  I don't have any moral objection to `assert` and don't feel like there's a need to draw a hard line between code idioms in "OG" versus "new" code. What all that in mind, I think @fabiangreffrath 's latest suggestion is good.

**Comment by fabiangreffrath** [comment]
Just done this for Woof!:
https://github.com/fabiangreffrath/woof/commit/c7b2362a590e5ee07d535c6ad8d955e7443bdca3

File: src/doom/g_game.c [hunk 2]
```diff
@@ -17,6 +17,7 @@
 
 
 
+#include <assert.h>
```
Comment (fabiangreffrath) [inline src/doom/g_game.c:20]: We won't be this anymore.

Comment (turol) [reply src/doom/g_game.c]: Removed

**Comment by turol** [comment]
I don't think we should remove cppcheck. I looked through the logs and it has found some leaks and questionable code.

**Review by fabiangreffrath — APPROVED** [review]
Thanks!

**Pull request merged**